### PR TITLE
Enabling SourceLink support.

### PR DESCRIPTION
**Pull request by dannyvv**
This change adds the nuget dependencies for sourcelink. What these packages do is adding a few targets that run during the build and produce a `myproj.sourcelink.json` file with something like:
```json
{"documents":{"Z:\\src\\react-native-windows\\*":"https://raw.githubusercontent.com/microsoft/react-native-windows/447f92738afcf6b3ccc567419455dd562f67c387/*"}}
```
which is then passed to all tools like cl.exe, link.exe, csc.exe etc and they will enrich the PDB with this information so that when this PDB is used on a machine without an enlistment that the file is downloaded from github.
See [SourceLink](https://github.com/dotnet/sourcelink/) for more technical details

It is unfortunate that UWP `.vcxproj` does not support `PackageRefernce` yet and is still on the old packages.config with ugly msbuild in the project needed for VisualSTudio to managed the nuget pacakges through the UI. So therefore the change is rather ugly in that it can't be centralized in `build.directory.props`

closing #5532 

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5533)

### CODE REVIEW

File: vnext/Desktop.DLL/React.Windows.Desktop.DLL.vcxproj [hunk 1]
```diff
@@ -1,5 +1,8 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="16.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\packages\Microsoft.SourceLink.GitHub.1.0.0\build\Microsoft.SourceLink.GitHub.props" Condition="Exists('..\packages\Microsoft.SourceLink.GitHub.1.0.0\build\Microsoft.SourceLink.GitHub.props')" />
```
Comment (dannyvv) [inline vnext/Desktop.DLL/React.Windows.Desktop.DLL.vcxproj:3]: This *gunk* is maintained by visual studio to handle the packages. When UWP upgrades to modern msbuild and supports Packagereference, all this can be removed and we can add the packagereferences in the shared `build.directory.props` file.

File: vnext/Desktop/React.Windows.Desktop.vcxproj.filters [hunk 2]
```diff
@@ -84,9 +84,6 @@
     <Midl Include="ABI\TestController.idl">
       <Filter>ABI</Filter>
     </Midl>
-    <Midl Include="..\Microsoft.ReactNative\ReactInstance.idl">
```
Comment (dannyvv) [inline vnext/Desktop/React.Windows.Desktop.vcxproj.filters:87]: caused by roundtripping in VS.

File: vnext/Desktop.DLL/React.Windows.Desktop.DLL.vcxproj (hunk 1 above)
Comment (dannyvv) [reply vnext/Desktop.DLL/React.Windows.Desktop.DLL.vcxproj:3]: Note: had to make a small update to the vs generated stuff to handle customers building our code.

File: change/react-native-windows-2020-07-15-21-24-35-master.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "prerelease",
+  "comment": "Enabeling SourceLink support.",
```
Comment (vmoroz) [inline change/react-native-windows-2020-07-15-21-24-35-master.json:3]: >Enabeling [](start = 14, length = 9)

nit: spelling

File: vnext/Desktop/React.Windows.Desktop.vcxproj [hunk 4]
```diff
@@ -1,5 +1,8 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="16.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\packages\Microsoft.SourceLink.GitHub.1.0.0\build\Microsoft.SourceLink.GitHub.props" Condition="Exists('..\packages\Microsoft.SourceLink.GitHub.1.0.0\build\Microsoft.SourceLink.GitHub.props')" />
```
Comment (vmoroz) [inline vnext/Desktop/React.Windows.Desktop.vcxproj:3]: >Import [](start = 3, length = 6)

Please move the new imports down where we import the C++/WinRT props file.
Having it here prevents us from using the $(SolutionDir) variable.

File: vnext/Desktop/React.Windows.Desktop.vcxproj [hunk 5]
```diff
@@ -314,6 +317,9 @@
     <Import Project="$(SolutionDir)packages\Microsoft.Windows.CppWinRT.2.0.200615.7\build\native\Microsoft.Windows.CppWinRT.targets" Condition="Exists('$(SolutionDir)packages\Microsoft.Windows.CppWinRT.2.0.200615.7\build\native\Microsoft.Windows.CppWinRT.targets')" />
     <Import Project="$(V8_Package)\build\native\ReactNative.V8JSI.Windows.targets" Condition="Exists('$(V8_Package)\build\native\ReactNative.V8JSI.Windows.targets') AND '$(USE_V8)' == 'true'" />
     <Import Project="$(SolutionDir)packages\ReactNative.Hermes.Windows.0.1.6\build\native\ReactNative.Hermes.Windows.targets" Condition="Exists('$(SolutionDir)packages\ReactNative.Hermes.Windows.0.1.6\build\native\ReactNative.Hermes.Windows.targets')" />
+    <Import Project="..\packages\Microsoft.Build.Tasks.Git.1.0.0\build\Microsoft.Build.Tasks.Git.targets" Condition="Exists('..\packages\Microsoft.Build.Tasks.Git.1.0.0\build\Microsoft.Build.Tasks.Git.targets')" />
```
Comment (vmoroz) [inline vnext/Desktop/React.Windows.Desktop.vcxproj:320]: >.. [](start = 21, length = 2)

Please use the $(SolutionDir)

File: vnext/Directory.Build.props [hunk 6]
```diff
@@ -11,6 +11,12 @@
 
     <!-- This property sets the CscTask to print the full path of the csharp file on errors and warnings rather than just the filename. -->
     <GenerateFullPaths>true</GenerateFullPaths>
+
+    <!-- 
+      SourceLink is disabled by default since customer projects are still building our projects and therefore we don't want to force customers to use this.
+      The PR (windows-vs-pr.yml) and CI (publish.yml() turn it back on.
+    -->
+    <EnableSourceLink>false</EnableSourceLink>
```
Comment (vmoroz) [inline vnext/Directory.Build.props:19]: >EnableSourceLink [](start = 5, length = 16)

Could we make it conditional if it is not set yet?

File: vnext/Desktop/React.Windows.Desktop.vcxproj.filters (hunk 2 above)
Comment (vmoroz) [reply vnext/Desktop/React.Windows.Desktop.vcxproj.filters:87]: Could you undo it? VS often does strange thing to out filters files. I often must restart VS just because of it.

---
In reply to: [455502717](https://github.com/microsoft/react-native-windows/pull/5533#discussion_r455502717) [](ancestors = 455502717)

File: vnext/Desktop/React.Windows.Desktop.vcxproj (hunk 4 above)
Comment (vmoroz) [reply vnext/Desktop/React.Windows.Desktop.vcxproj:3]: The same is for other project files.

---
In reply to: [456013669](https://github.com/microsoft/react-native-windows/pull/5533#discussion_r456013669) [](ancestors = 456013669)

File: vnext/Desktop/React.Windows.Desktop.vcxproj (hunk 5 above)
Comment (vmoroz) [reply vnext/Desktop/React.Windows.Desktop.vcxproj:320]: The same is for other project files.

---
In reply to: [456013866](https://github.com/microsoft/react-native-windows/pull/5533#discussion_r456013866) [](ancestors = 456013866)

File: vnext/JSI/Desktop/JSI.Desktop.vcxproj [hunk 7]
```diff
@@ -1,5 +1,8 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="16.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (vmoroz) [inline vnext/JSI/Desktop/JSI.Desktop.vcxproj:2]: >Project [](start = 1, length = 7)

I wonder why these changes are needed here and not in other projects. This project does not create a DLL. We have several such projects. The most important among them are ReactCommon and Folly.

File: vnext/Desktop/React.Windows.Desktop.vcxproj (hunk 5 above)
Comment (dannyvv) [reply vnext/Desktop/React.Windows.Desktop.vcxproj:320]: Sigh....  @microsoft/vcpp I hope this will be fixed in the next version of msvc/vs/msbuild..

File: vnext/Directory.Build.props (hunk 6 above)
Comment (dannyvv) [reply vnext/Directory.Build.props:19]: If you specify it as a global property, msbuild will not allow it to be set ever, so when building with `/p:SourceLink=true` setting it to false will be ignored. 
I'll add the condition to be safe for customer projects, but this is generally imported early enough for any customer project to not be able to set it themselves.

File: vnext/JSI/Desktop/JSI.Desktop.vcxproj (hunk 7 above)
Comment (dannyvv) [reply vnext/JSI/Desktop/JSI.Desktop.vcxproj:2]: I did all projects that have a package.config and are not tests. I'll remove it form this one then.

File: vnext/Desktop/React.Windows.Desktop.vcxproj (hunk 5 above)
Comment (dannyvv) [reply vnext/Desktop/React.Windows.Desktop.vcxproj:320]: Note that I'll have to move the .props usages down then as well since SolutionDir is only created after `Microsfot.Cpp.Defaults.Props`

Comment (dannyvv) [reply vnext/Desktop/React.Windows.Desktop.vcxproj:320]: Note we have been pretty inconsistent on SolutionDir. I'll send another PR to fix up the rest...

File: vnext/Desktop/React.Windows.Desktop.vcxproj (hunk 4 above)
Comment (dannyvv) [reply vnext/Desktop/React.Windows.Desktop.vcxproj]: yea, did so as part of solutiondir change.
